pkg/asset/agent/mirror/registriesconf.go

```diff
@@ -139,14 +140,24 @@ func (i *RegistriesConf) Generate(dependencies asset.Parents) error {
 	releaseImage := &releaseimage.Image{}
 	dependencies.Get(installConfig, releaseImage)
 
-	if !installConfig.Supplied || len(installConfig.Config.ImageContentSources) == 0 {
+	if !installConfig.Supplied || len(installConfig.Config.ImageContentSources) == 0 && len(installConfig.Config.ImageDigestSources) == 0 {
 		return i.generateDefaultRegistriesConf()
 	}
 
+	if len(installConfig.Config.ImageContentSources) != 0 && len(installConfig.Config.ImageDigestSources) != 0 {
+		return fmt.Errorf("invalid install-config.yaml, cannot set imageContentSources and imageDigestSources at the same time")
+	}
+
 	registries := &sysregistriesv2.V2RegistriesConf{
 		Registries: []sysregistriesv2.Registry{},
 	}
-	for _, group := range bootstrap.MergedMirrorSets(installConfig.Config.ImageContentSources) {
+	digestMirrorSources := []types.ImageDigestSource{}
+	if len(installConfig.Config.ImageContentSources) > 0 {
+		digestMirrorSources = bootstrap.ContentSourceToDigestMirror(installConfig.Config.ImageContentSources)
```

Comment:
If this conversion were done in `ConvertInstallConfig()`, it would result in the cluster always getting an `ImageDigestMirrorSet` CR instead of an `ImageContentSourcePolicy` regardless of which field the user specified in the install-config. The decision to convert only here is presumably to avoid that.

In the agent installer, assisted-service generates a (second) install-config and it currently sets the `imageContentSources` field. This conversion means that we have lost any information as to whether the user specified `imageContentSources` or `imageDigestSources`, which I think is probably OK (on line 167 we set `MirrorByDigestOnly = true` anyway, so effectively we have always required digest-only mirrors in the agent installer). However, it will take an additional change to assisted-service to switch it to always setting `imageDigestSources` for `ImageDigestMirrorSet` CRs to start showing up in clusters deployed with the agent installer. Without that, it will always create `ImageContentSourcePolicy` CRs even if the user specifies `imageDigestSources` in the install-config.

That does make me wonder whether we could afford to always create ImageDigestMirrorSet` CRs on every type of deployment regardless of which field the user specifies (which wouldn't need a change in assisted-service). But I'm sure you've already spent a lot more time thinking about it than I.

Reply:
I take a rough look at the assisted-service,  if we always create ImageDigestMirrorSet, the logic to generate install-config can stay the same, only the code writes the `ImageContentSourcePolicy` manifest needs to update.

@bparees @patrickdillon What do you think about only creating `ImageDigestMirrorSet` CRs on every type of deployment regardless of which field the user specifies? If we do it this way, adding the new field seems not necessary.

Reply:
FYI I opened [MGMT-14159](https://issues.redhat.com/browse/MGMT-14159) in case any adjustments need to be made in assisted installer.

Reply:
@avishayt thanks. I linked that card to the epic that tracks a set of ICSP migrations.

Reply:
Does the assisted-service generate the install-config usong data from this registries.conf? So assisted service can not know if the mirror registry is from `ImageContentSources` or `ImageDigestSources` of the first install-config?

Reply:
@bfournie implemented most of the code in assisted-service that deals with ImageContentSources.  Bob, can you comment?

Reply:
@avishayt yes, I will make any changes to the assisted-service for this. However it looks like with the most recent change to use `ContentSourceToDigestMirror` that the ImageContentSources are converted to ImageDigestSources by the installer so no changes would be required in assisted-service

Reply:
@bfournie  That convert function can ImageContent -> ImageDigest, and the converted contented will be written to `RegistriesConf` struct. Does this help? https://github.com/openshift/installer/pull/6235/files#diff-beb52b8fa67370e9332766c49cad67ee11cfe31bd15fdbbc7bf258bfdeadf632R156
or you mean the converted data should change the first install-config.yaml passed to the installer?

Reply:
@bfournie as far as I can see it is still only translating at the point they're used to create the ignition file; the manifests are still generated based on the field actually set in the install-config.
To change that, the translation would have to happen in 'pkg/types/conversion/install-config.go`, so that everything sees the install-config as if the user had specified the new fields.